File: scripts/test_solutions/994_rotting_oranges.py
from collections import deque

def orangesRotting(grid):
    m, n = len(grid), len(grid[0])
    queue = deque()
    fresh = 0

    for i in range(m):
        for j in range(n):
            if grid[i][j] == 2:
                queue.append((i, j))
            elif grid[i][j] == 1:
                fresh += 1

    if fresh == 0:
        return 0

    dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    minutes = 0

    while queue:
        size = len(queue)
        for _ in range(size):
            r, c = queue.popleft()
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if 0 <= nr < m and 0 <= nc < n and grid[nr][nc] == 1:
                    grid[nr][nc] = 2
                    fresh -= 1
                    queue.append((nr, nc))
        if queue:
            minutes += 1

    return minutes if fresh == 0 else -1

File: scripts/test_solutions/994_rotting_oranges.js
var orangesRotting = function(grid) {
    const m = grid.length, n = grid[0].length, queue = [];
    let fresh = 0;
    for (let i = 0; i < m; i++) for (let j = 0; j < n; j++) {
        if (grid[i][j] === 2) queue.push([i, j]);
        else if (grid[i][j] === 1) fresh++;
    }
    if (fresh === 0) return 0;
    const dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]];
    let minutes = 0;
    while (queue.length) {
        const size = queue.length;
        for (let i = 0; i < size; i++) {
            const [r, c] = queue.shift();
            for (const [dr, dc] of dirs) {
                const nr = r + dr, nc = c + dc;
                if (nr >= 0 && nr < m && nc >= 0 && nc < n && grid[nr][nc] === 1) {
                    grid[nr][nc] = 2; fresh--; queue.push([nr, nc]);
                }
            }
        }
        if (queue.length) minutes++;
    }
    return fresh === 0 ? minutes : -1;
};
